spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java

```diff
@@ -0,0 +1,183 @@
+            return false;
+        }
+
+        boolean verify(String dateFormat) {
+            if (!containsAny(dateFormat, this.keywords)
+                    || (this.ignore != null && dateFormat.contains(this.ignore))) {
+                return false;
+            }
+
+            boolean cond = containsAny(dateFormat, this.flags);
+            if (this.isRequired) {
+                return !cond;
+            }
+            return cond;
+        }
+    }
+
+    public DateFormatStringChecker(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        if (!(seen == Const.INVOKESPECIAL || seen == Const.INVOKEVIRTUAL || seen == Const.INVOKESTATIC
+                || seen == Const.INVOKEINTERFACE) || stack.getStackDepth() == 0) {
+            return;
+        }
+
+        int idx = -1;
+        String cl = getClassConstantOperand();
+        String nm = getNameConstantOperand();
+        String si = getSigConstantOperand();
+        if ("java/text/SimpleDateFormat".equals(cl)
+                && contains(nm, Const.CONSTRUCTOR_NAME, "applyPattern", "applyLocalizedPattern")) {
+            switch (si) {
+            case "(Ljava/lang/String;)V":
+                idx = 0;
+                break;
+            case "(Ljava/lang/String;Ljava/util/Locale;)V":
+            case "(Ljava/lang/String;Ljava/text/DateFormatSymbols;)V":
+                idx = 1;
+                break;
+            }
+        } else if ("java/time/format/DateTimeFormatter".equals(cl) && "ofPattern".equals(nm)) {
+            switch (si) {
+            case "(Ljava/lang/String;)Ljava/time/format/DateTimeFormatter;":
+                idx = 0;
+                break;
+            case "(Ljava/lang/String;Ljava/util/Locale;)Ljava/time/format/DateTimeFormatter;":
+                idx = 1;
+                break;
+            }
+        } else if ("org/apache/commons/lang3/time/FastDateFormat".equals(cl) && "getInstance".equals(nm)) {
+            switch (si) {
+            case "(Ljava/lang/String;)Lorg/apache/commons/lang3/time/FastDateFormat;":
+                idx = 0;
+                break;
+            case "(Ljava/lang/String;Ljava/util/Locale;)Lorg/apache/commons/lang3/time/FastDateFormat;":
+            case "(Ljava/lang/String;Ljava/util/TimeZone;)Lorg/apache/commons/lang3/time/FastDateFormat;":
+                idx = 1;
+                break;
+            }
+        }
+
+        if (idx == -1) {
+            return;
+        }
+
+        String dateFormatString = (String) stack.getStackItem(idx).getConstant();
+        if (dateFormatString != null && runDateFormatRuleVerify(dateFormatString)) {
+            bugReporter.reportBug(new BugInstance(this, "FS_BAD_DATE_FORMAT_FLAG_COMBO", NORMAL_PRIORITY)
+                    .addClassAndMethod(this).addCalledMethod(this).addString(dateFormatString)
+                    .describe(StringAnnotation.FORMAT_STRING_ROLE).addSourceLine(this));
```

Comment:
```suggestion
            bugReporter.reportBug(new BugInstance(this, "FS_BAD_DATE_FORMAT_FLAG_COMBO", NORMAL_PRIORITY)
                    .addClassAndMethod(this)
                    .addCalledMethod(this)
                    .addString(dateFormatString).describe(StringAnnotation.FORMAT_STRING_ROLE)
                    .addSourceLine(this));
```
Can you please reformat it similarly so it's easy to see what belongs together?

Reply:
fixed